src/NHibernate/Loader/JoinWalker.cs

```diff
@@ -299,26 +312,48 @@ private void AddAssociation(OuterJoinableAssociation association)
 		/// </summary>
 		protected void WalkEntityTree(IOuterJoinLoadable persister, string alias)
 		{
-			WalkEntityTree(persister, alias, string.Empty, 0);
+			WalkEntityTree(persister, alias, String.Empty);
+			ProcessJoins();
 		}
 
 		/// <summary>
 		/// For a collection role, return a list of associations to be fetched by outerjoin
 		/// </summary>
 		protected void WalkCollectionTree(IQueryableCollection persister, string alias)
 		{
-			WalkCollectionTree(persister, alias, string.Empty, string.Empty, 0);
-			//TODO: when this is the entry point, we should use an INNER_JOIN for fetching the many-to-many elements!
+			WalkCollectionTree(persister, alias, String.Empty, String.Empty);
+			ProcessJoins();
+		}
+
+		protected void ProcessJoins()
+		{
+			while(joinQueue.Count > 0)
+			{
+				QueueEntry entry = joinQueue.Dequeue();
+
+				switch(entry)
+				{
+					case CollectionQueueEntry ce:
+						WalkCollectionTree(ce.Persister, ce.Alias, ce.Path, ce.PathAlias);
+						break;
+					case EntityQueueEntry eqe:
+						WalkEntityTree(eqe.Persister, eqe.Alias, eqe.Path);
+						break;
+					default:
```

Comment:
`default` is never reached here. The `QueueEntry` is always either `CollectionQueueEntry` or `EntityQueueEntry`.

Reply:
Oh, I see... you enqueue `null` there.